### ISSUE TITLE
OIDC and Refresh Tokens

### ISSUE THREAD

**Issue by fabiosimeoni** [FusionAuth/fusionauth-issues#110]
I must be doing something wrong to refresh id tokens in an OIDC dance. I have:

- enabled refresh tokens for my app...
- used `scope=openid%20offline_access` when redirecting to the `authorize` endpoint... 
- got back both `id_token` and `refresh_token` when I later resolve the access code with the `token` endpoint...
- hit again the `token` endpoint with the `refresh_token` when my `id_token` expires...
- got back just an `access_token` when I do, no new `id_token`..why?

what am I missing? thanks. 

(and thank you for the good work on FusionAuth, getting better by the day!)

**Comment by robotdan** [comment]
Hi @fabiosimeoni, thanks for the compliment! 

Hmm... the `id_token` should be returned.  I have a bunch of tests for this scenario, perhaps we are missing an edge case. 

When you use the Refresh Grant, are you sending in the `scope` parameter? If you omit it - we'll use the scopes requested during the authorization request. If you do offer it, it must match what was initially requested. 

That is the only thing I can think of - but you should be seeing any error if you send in scopes that are not equal to the original. 

Can you show me a rough example of your authorization request and refresh grant request? Thanks.

**Comment by fabiosimeoni** [comment]
Hi @robotdan, thanks for the quick reply.

Initially I didn’t use the scope parameter because it didn't find it in the documentation, then I tried it to force it with `openid` first, then `openid%20offline_access`. In both cases it would indeed tell me they didn’t match the original. As omitting the scope yields only an access token, it’s as if the original scope got lost in the dance, and the default kicked in. 


`authorize` request:

```
GET https://localhost:5443/oauth2/authorize?state=be/logged&client_id=ef9626ad-84ce-4dc4-a64d-6f420bec38fb&redirect_uri=https://localhost:5443/be/oauth2/callback&response_type=code&scope=offline_access%20openid
```

`token` request/response (Authorization Grant) :

```
POST http://auth:9011/oauth2/token?code=bPREvfObnmyTV74ejRlvO_cwxiC69XuSOgW9HnCL2L4&client_id=ef9626ad-84ce-4dc4-a64d-6f420bec38fb&grant_type=authorization_code&redirect_uri=https://localhost:5443/be/oauth2/callback

200
{"access_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImVmOTYyNmFkLTg0Y2UtNGRjNC1hNjRkLTZmNDIwYmVjMzhmYiJ9.eyJhdWQiOiJlZjk2MjZhZC04NGNlLTRkYzQtYTY0ZC02ZjQyMGJlYzM4ZmIiLCJleHAiOjE1NTQ4NDM1MDYsImlhdCI6MTU1NDg0MzUwMSwiaXNzIjoiYWNtZS5jb20iLCJzdWIiOiJkYzBlMDQxZC1hMTYyLTQ5MDgtODU3NS05YmNiZjExOWExODMiLCJhdXRoZW50aWNhdGlvblR5cGUiOiJQQVNTV09SRCIsImVtYWlsIjoiam9obmRvZUBhcHByaXNlY3JhZnQuY29tIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsInByZWZlcnJlZF91c2VybmFtZSI6ImpvaG5kb2UiLCJhcHBsaWNhdGlvbklkIjoiZWY5NjI2YWQtODRjZS00ZGM0LWE2NGQtNmY0MjBiZWMzOGZiIiwicm9sZXMiOltdfQ.fV3x5sGMMRQoH4nXvhE_ilxZesYCgGiTvyrL-SMPUvE","expires_in":4,"id_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImVmOTYyNmFkLTg0Y2UtNGRjNC1hNjRkLTZmNDIwYmVjMzhmYiJ9.eyJhdWQiOiJlZjk2MjZhZC04NGNlLTRkYzQtYTY0ZC02ZjQyMGJlYzM4ZmIiLCJleHAiOjE1NTQ4NDM1MDYsImlhdCI6MTU1NDg0MzUwMSwiaXNzIjoiYWNtZS5jb20iLCJzdWIiOiJkYzBlMDQxZC1hMTYyLTQ5MDgtODU3NS05YmNiZjExOWExODMiLCJhdXRoZW50aWNhdGlvblR5cGUiOiJQQVNTV09SRCIsImVtYWlsIjoiam9obmRvZUBhcHByaXNlY3JhZnQuY29tIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsInByZWZlcnJlZF91c2VybmFtZSI6ImpvaG5kb2UiLCJhdF9oYXNoIjoiV0hQeEttajhReUNHSlk1VlV6Vl9LQSIsImNfaGFzaCI6IlZVemdfSGZ5aTdid2UyVURqdWhMc0EiLCJhcHBsaWNhdGlvbklkIjoiZWY5NjI2YWQtODRjZS00ZGM0LWE2NGQtNmY0MjBiZWMzOGZiIiwicm9sZXMiOltdfQ.vB0iIMHHFm5p8gFphnqwjq6UC7POWutrhOvtvrE1Ur0","refresh_token":"s5wpHUj5yq1x-ffF1nRp6PKJavT2Rwfmpt_M-Lz6PK1FJ47bBP5nUA","token_type":"Bearer","userId":"dc0e041d-a162-4908-8575-9bcbf119a183"}
```
`token` request/response (Refresh Grant) :

```
POST http://auth:9011/oauth2/token?client_id=ef9626ad-84ce-4dc4-a64d-6f420bec38fb&grant_type=refresh_token&refresh_token=s5wpHUj5yq1x-ffF1nRp6PKJavT2Rwfmpt_M-Lz6PK1FJ47bBP5nUA&scope=offline_access%20openid

200
{"access_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImVmOTYyNmFkLTg0Y2UtNGRjNC1hNjRkLTZmNDIwYmVjMzhmYiJ9.eyJhdWQiOiJlZjk2MjZhZC04NGNlLTRkYzQtYTY0ZC02ZjQyMGJlYzM4ZmIiLCJleHAiOjE1NTQ4NDkwODcsImlhdCI6MTU1NDg0OTA4MiwiaXNzIjoiYWNtZS5jb20iLCJzdWIiOiJkYzBlMDQxZC1hMTYyLTQ5MDgtODU3NS05YmNiZjExOWExODMiLCJhdXRoZW50aWNhdGlvblR5cGUiOiJSRUZSRVNIX1RPS0VOIiwiZW1haWwiOiJqb2huZG9lQGFwcHJpc2VjcmFmdC5jb20iLCJlbWFpbF92ZXJpZmllZCI6dHJ1ZSwicHJlZmVycmVkX3VzZXJuYW1lIjoiam9obmRvZSIsInJvbGVzIjpbXSwiYXBwbGljYXRpb25JZCI6ImVmOTYyNmFkLTg0Y2UtNGRjNC1hNjRkLTZmNDIwYmVjMzhmYiJ9.u02Lmv9qpePmkfIaZlte6bgPgL9nj4Vo6ZAqnCHZX1E","expires_in":5,"token_type":"Bearer","userId":"dc0e041d-a162-4908-8575-9bcbf119a183"}
```

Note the aggressive expiration times are just to drive testing. So my tokens last for 5 seconds and my refresh tokens last for 15. Ultimately, I'd like to exercise the cycle: login-use-refresh-use-.

**Comment by robotdan** [comment]
Thanks for that additional info, I'll see if I can recreate. 

Also - you're correct, looks like `scope` is missing from the doc for the Refresh Grant request, I'll fix that. Thanks for pointing that out.

**Comment by robotdan** [comment]
Found it, was able to recreate. This will be fixed in the upcoming release. Thanks!

**Comment by robotdan** [comment]
Interesting, reading through different issues it seems there are several perspectives on returning the `id_token` on the Refresh Grant.  In case anyone is interested in different opinions, continue reading.

&nbsp;
The [OpenID Connect Core 1.0 spec Section 12.2 Successful Refresh Response](https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse) says:

> Upon successful validation of the Refresh Token, the response body is the Token Response of Section 3.1.3.3 except that it might not contain an id_token.

&nbsp;
This issue for IdentityServer has a user asking for the `id_token` on the Refresh response, and it doesn't sound like they'll add it.  (Jan 10, 2017)
- https://github.com/IdentityServer/IdentityServer3/issues/3447

&nbsp;
Just a few months later, this issue for IdentityServer sounds like they started returning an `id_token` and users didn't want it, or were at least surprised by this addition. (Apr 5, 2017)
- https://github.com/IdentityServer/IdentityServer3/issues/3621

&nbsp;
This StackOverflow answer regarding Auth0 seems to think they you should never get an `id_token` during the Refresh Grant because it represents and authentication event. 
- https://stackoverflow.com/questions/42615253/auth0-get-id-token-from-refresh-token

&nbsp;
This issue for a Google dotnet client indicates that they do not return the `id_token` either and it is working as designed. They provide some rationale in this comment:
- https://github.com/googleapis/google-api-dotnet-client/issues/1141#issuecomment-360239385

&nbsp;
In summary - it looks like everyone has an opinion. It seems that some integrations rely upon this token, or at least generally expect it. While the intention of the `id_token` as described by the specification doesn't necessarily make sense on the Refresh response, I don't see any harm in keeping it. 

&nbsp;
Perhaps if we think it is additional overhead, we could add a configuration at a later date to optionally return it.

**Comment by fabiosimeoni** [comment]
I’ve come across the opinion of an id token as evidence of a recent authentication event, hence of ‘user presence’. Making id tokens short-lived goes well with the idea, while refreshing might seem to go against it. (And yet it may be argued that if the token is refreshed as it is presented, no at an arbitrary point later, isn’t the user actually “there”, present, only with expired credentials?)

I’m not an expert at all here, but the idea that I can keep users happily logged — maybe even implement the token-equivalent of sliding sessions — and yet revoke rogue tokens (or simply tokens with stale info) at refresh time, seems appealing to me. Maybe the ambiguity in that excerpt of the specs relates to the different grants, eg. tokens would not be refreshed in grants other than Authorization Grant?

I can’t yet ‘see’ any reason why the logic/conveniences of refresh tokens wouldn’t apply to id tokens too. And the specs would have explicitly forbidden the id token from being the refresh grant response, no? 

I’d be interested to hear more expert opinions.

**Comment by robotdan** [comment]
> I can’t yet ‘see’ any reason why the logic/conveniences of refresh tokens wouldn’t apply to id tokens too. And the specs would have explicitly forbidden the id token from being the refresh grant response, no?

I think I agree with this - most of the arguments against it that I've read have more to do with performance, or the perceived lack of necessity rather than any technical or security related concern. 

Most specifications (including OIDC) leave many details up to interpretation or preference. Our guiding principle with FusionAuth is to err on the side of making the product more flexible and let the developer decide how they want to use the product.

**Comment by robotdan** [comment]
Fixed in 1.6.0, let me know if you still have issues with this one.
